lib/board.rb

```diff
@@ -24,4 +28,64 @@ def available?(input)
     space = values[input - 1]
     space.is_a?(Integer)
   end
+
+  def has_win?
+    outcome_checker.win?(self)
+  end
+
+  def has_draw?
+    outcome_checker.draw?(self)
+  end
+
+  def rows
+    values_grid
+  end
+
+  def columns
+    values_grid.transpose
+  end
+
+  def diagonals
+    [left_diagonal, right_diagonal]
+  end
+
+  private
+
+  def dimension
+    Math.sqrt(values.length).to_i
+  end
+
+  def values_grid
+    grid = []
+    values.each_slice(dimension) { |row| grid << row }
+    grid
+  end
+
+  def left_diagonal
+    left_diagonal = []
```

Comment:
with regards to shortening these diagonals methods: conceptually, you're pulling out items that match certain criteria (are at particular indices) so i'd start with the filtering methods that arrays already have in order to eliminate the explicit looping. 

that might even allow you to factor out the duplicated logic between the 2 since they're _almost_ identical.

Reply:
Good idea - thank you!